libminifi/src/core/logging/alert/AlertSink.cpp

```diff
@@ -0,0 +1,269 @@
+      (void)services.release();
+    }
+  }
+}
+
+AlertSink::~AlertSink() {
+  {
+    std::lock_guard lock(mtx_);
+    running_ = false;
+    cv_.notify_all();
+  }
+  if (flush_thread_.joinable()) {
+    flush_thread_.join();
+  }
+  delete services_.exchange(nullptr);
+}
+
+void AlertSink::send(Services& services) {
+  LogBuffer logs;
+  buffer_.commit();
+  if (!buffer_.tryDequeue(logs)) {
+    return;
+  }
+
+  auto client = core::ClassLoader::getDefaultClassLoader().instantiate<utils::BaseHTTPClient>("HTTPClient", "HTTPClient");
+  if (!client) {
+    logger_->log_error("Could not instantiate a HTTPClient object");
+    return;
+  }
+  client->initialize("PUT", config_.url, services.ssl_service);
+
+  rapidjson::Document doc(rapidjson::kObjectType);
+  std::string agent_id = services.agent_id->getAgentIdentifier();
+  doc.AddMember("agentId", rapidjson::Value(agent_id.data(), agent_id.length()), doc.GetAllocator());
+  doc.AddMember("alerts", rapidjson::Value(rapidjson::kArrayType), doc.GetAllocator());
+  for (const auto& [log, _] : logs.data_) {
+    doc["alerts"].PushBack(rapidjson::Value(log.data(), log.size()), doc.GetAllocator());
+  }
+  rapidjson::StringBuffer buffer;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(buffer);
+  doc.Accept(writer);
+
+  auto data_input = std::make_unique<utils::ByteInputCallback>();
+  auto data_cb = std::make_unique<utils::HTTPUploadCallback>();
+  data_input->write(std::string(buffer.GetString(), buffer.GetSize()));
+  data_cb->ptr = data_input.get();
+  client->setUploadCallback(data_cb.get());
+  client->setContentType("application/json");
+
+  bool req_success = client->submit();
+
+  int64_t resp_code = client->getResponseCode();
+  const bool response_success = 200 <= resp_code && resp_code < 300;
+  const bool client_err = 400 <= resp_code && resp_code < 500;
+  const bool server_err = 500 <= resp_code && resp_code < 600;
+  if (client_err || server_err) {
+    logger_->log_error("Error response code '" "%" PRId64 "' from '%s'", resp_code, config_.url);
+  } else if (!response_success) {
+    logger_->log_warn("Non-success response code '" "%" PRId64 "' from '%s'", resp_code, config_.url);
+  } else {
+    logger_->log_debug("Response code '" "%" PRId64 "' from '%s'", resp_code, config_.url);
+  }
+
+  if (!req_success) {
+    logger_->log_error("Failed to send alert request");
+  }
+}
+
+AlertSink::LogBuffer AlertSink::LogBuffer::allocate(size_t /*size*/) {
+  return {};
+}
+
+AlertSink::LogBuffer AlertSink::LogBuffer::commit() {
+  return std::move(*this);
+}
+
+size_t AlertSink::LogBuffer::size() const {
+  return size_;
+}
+
+void AlertSink::LiveLogSet::setLifetime(std::chrono::milliseconds lifetime) {
+  lifetime_ = lifetime;
+}
+
+bool AlertSink::LiveLogSet::tryAdd(std::chrono::milliseconds now, size_t hash) {
+  auto limit = now - lifetime_;
+  while (!timestamped_hashes_.empty() && timestamped_hashes_.front().first < limit) {
```

Comment:
This is assuming that `timestamped_hashes_` is having the hashes ordered in a chronological order. I think it would be better to use a priority queue for that. Also, they could become invalid if `setLifeTime` is called between two `tryAdd`s and `lifetime_` is modified. But since I just see one call for it in the ` AlertSink` constructor, can this method be removed and `lifetime_` set from constructor only?

Reply:
I don't remember what the purpose of `setLifeTime` was, removed it

Reply:
Okay. Wouldn't you think a priority queue would better fit the role of `timestamped_hashes_`, so that it is always in a chronological order? Don't do it if it's too complicated, just an idea.

Reply:
I think chronological order is already guaranteed as by default we use a steady clock